syncer/syncer.go

```diff
@@ -1828,103 +1757,14 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext) e
 		return err
 	}
 
-	if s.cfg.ShardMode == "" {
-		s.tctx.L().Info("start to handle ddls in normal mode", zap.String("event", "query"), zap.Strings("ddls", needHandleDDLs), zap.ByteString("raw statement", ev.Query), log.WrapStringerField("location", ec.currentLocation))
-
-		// interrupted after flush old checkpoint and before track DDL.
-		failpoint.Inject("FlushCheckpointStage", func(val failpoint.Value) {
-			err = handleFlushCheckpointStage(1, val.(int), "before track DDL")
-			if err != nil {
-				failpoint.Return(err)
-			}
-		})
-
-		// run trackDDL before add ddl job to make sure checkpoint can be flushed
-		for _, td := range needTrackDDLs {
-			if err = s.trackDDL(usedSchema, td.rawSQL, td.tableNames, td.stmt, &ec); err != nil {
-				return err
-			}
-		}
-
-		// interrupted after track DDL and before execute DDL.
-		failpoint.Inject("FlushCheckpointStage", func(val failpoint.Value) {
-			err = handleFlushCheckpointStage(2, val.(int), "before execute DDL")
-			if err != nil {
-				failpoint.Return(err)
-			}
-		})
-
-		job := newDDLJob(nil, needHandleDDLs, *ec.lastLocation, *ec.startLocation, *ec.currentLocation, *ec.traceID, sourceTbls)
-		err = s.addJobFunc(job)
-		if err != nil {
-			return err
-		}
-
-		// when add ddl job, will execute ddl and then flush checkpoint.
-		// if execute ddl failed, the execErrorDetected will be true.
-		err = s.execError.Get()
-		if err != nil {
-			return terror.ErrSyncerUnitHandleDDLFailed.Delegate(err, ev.Query)
-		}
-
-		s.tctx.L().Info("finish to handle ddls in normal mode", zap.String("event", "query"), zap.Strings("ddls", needHandleDDLs), zap.ByteString("raw statement", ev.Query), log.WrapStringerField("location", ec.currentLocation))
-
-		for _, table := range onlineDDLTableNames {
-			s.tctx.L().Info("finish online ddl and clear online ddl metadata in normal mode", zap.String("event", "query"), zap.Strings("ddls", needHandleDDLs), zap.ByteString("raw statement", ev.Query), zap.String("schema", table.Schema), zap.String("table", table.Name))
-			err = s.onlineDDL.Finish(ec.tctx, table.Schema, table.Name)
-			if err != nil {
-				return terror.Annotatef(err, "finish online ddl on %s.%s", table.Schema, table.Name)
-			}
-		}
-
-		return nil
-	}
-
-	// handle shard DDL in optimistic mode.
+	// handle shard DDL
 	if s.cfg.ShardMode == config.ShardOptimistic {
 		return s.handleQueryEventOptimistic(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames)
+	} else if s.cfg.ShardMode == config.ShardPessimistic {
+		return s.handleQueryEventPessimistic(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, ddlInfo, sqls)
 	}
```

Comment:
Also refine the normal mode？😂

Reply:
done in https://github.com/pingcap/dm/pull/896/commits/575352625260a2b8ae0e489d24c4b09a37dd407a